implementations/rust/router/src/router.rs

```diff
@@ -0,0 +1,168 @@
+#![allow(unused)]
+pub mod router {
+    use ockam_message::message::*;
+    use std::collections::HashMap;
+    use std::io::{Error, ErrorKind, Read, Write};
+    use std::net::UdpSocket;
+    use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
+    use std::ops::Add;
+    use std::sync::{Arc, Mutex};
+    use std::thread;
+
+    pub struct UdpAddressHandler {
+        pub socket: UdpSocket,
+    }
+
+    impl HandleMessage for UdpAddressHandler {
+        fn handle_message(&self, mut m: Box<Message>) -> Result<(), std::io::Error> {
+            println!("In UdpAddressHandler!");
+            Ok(())
+        }
+    }
+
+    pub trait HandleMessage {
+        fn handle_message(&self, m: Box<Message>) -> Result<(), std::io::Error>;
+    }
+
+    pub struct Router {
+        registry: HashMap<u64, Arc<Mutex<Box<dyn HandleMessage + Send>>>>,
+    }
+
+    impl Router {
+        pub fn new() -> Router {
+            Router {
+                registry: HashMap::new(),
+            }
+        }
+
+        pub fn register_handler(
+            &mut self,
+            handler: Arc<Mutex<Box<dyn HandleMessage + Send>>>,
+            address: &Address,
+        ) -> Result<(), String> {
+            let k = address.get_key()?;
+            self.registry.insert(k, handler);
+            Ok(())
+        }
+
+        pub fn route(&mut self, mut m: Box<Message>) -> Result<(), String> {
+            // Pop the first address in the list
+            // If there are no addresses, route to the controller
+            // Controller key is always 0
+            let mut key: u64 = 0;
+            let address: Address;
+            if !m.onward_route.addresses.is_empty() {
+                address = m.onward_route.addresses.remove(0);
```

Comment:
I think the router shouldn't remove the address.
It just hands over the message to a handler based on the `address type` of the first address in the `onward_route` of the message.

Reply:
So the handler removes the address? I think the router should be the only thing messing with the routes. All route management done by the router.